XenModel/Network/Heartbeat.cs

```diff
@@ -148,6 +148,23 @@ private void DoHeartbeat()
                 HandleConnectionLoss();
                 return;
             }
+            catch (WebException exn)
+            {
+                log.Error(exn);
+                if (((HttpWebResponse)exn.Response).StatusCode == HttpStatusCode.ProxyAuthenticationRequired) // work-around for CA-214653
+                {
+                    log.DebugFormat("Heartbeat for {0} has failed due to {1} credentials; closing the main connection",
+                        session == null ? "null" : session.Url,
+                        session.proxy.Proxy.Credentials == null ? "missing" : "incorrect");

```

Comment:
In this particular case it doesn't make much sense in english either to say "heartbeat for null has failed". Back to my earlier snippet I would write, for example:
```
if (session ==null) log.Debug("cannot find session") //or "session is null" etc.
else if (session.proxy.Proxy.Credentials == null) log.DebugFormat("heartbeat for {0} failed due to missing credentials", sessionurl);
else log.DebugFormat("heartbeat for {0} failed due to incorrect credentials", sessionurl);
```